src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,212 @@
+members on various JS types to enable this conversion, usually labeled as or
+prefixed with `toDart`. They are often marked `external`, and their
+implementation is usually platform-dependent. Note that not all JS types have a
+conversion, and not all Dart types have a conversion.
+
+In general, the conversion table looks like the following:
+
+| JS type                             | Dart type                                |
+| ----------------------------------- | ---------------------------------------- |
+| `JSNumber`, `JSBoolean`, `JSString` | `num`, `int`, `double`, `bool`, `String` |
+| `JSExportedFunction`                | `Function`                               |
+| `JSArray<T extends JSAny?>`         | `List<T extends JSAny?>`                 |
+| `JSPromise<T extends JSAny?>`       | `Future<T extends JSAny?>`               |
+| Typed arrays e.g. `JSUint8Array`    | `dart:typed_data`                        |
+| `JSBoxedDartObject`                 | Opaque Dart value                        |
+
+**Important**: Conversions may have different costs depending on the compiler.
+Prefer to only convert when you need to. Furthermore, conversions may or may not
+produce a new value. This doesn’t matter for immutable values like numbers, but
+does matter for types like `List`s. A conversion to a `JSArray` may produce a
+new value by copying or may not, so do not rely on modifications to the
+`JSArray` to affect the `List`. Typed array conversions have a similar
+limitation. Look at the specific conversion function for more details.
+
+## Requirements on `external` declarations and `Function.toJS`
+
+In order to ensure type safety and consistency, the compiler places requirements
+on what types can flow into and out of JS. Passing arbitrary Dart values into JS
+is not allowed. Instead, the compiler requires users to use a compatible interop
+type like a JS type or a primitive, which would then be implicitly converted by
+the compiler. For example, these would be allowed:
+
+{:.good}
+```dart
+@JS()
+external void primitives(String a, int b, double c, num d, bool e);
+```
+
+{:.good}
+```dart
+@JS()
+external JSArray jsTypes(JSObject _, JSString __);
+```
+
+{:.good}
+```dart
+extension type InteropType(JSObject _) {}
+
+@JS()
+external InteropType get interopType;
+```
+
+whereas these would return an error:
+
+{:.bad}
+```dart
+@JS()
+external Function get function;
+```
+
+{:.bad}
+```dart
+@JS()
+external set list(List _);
+```
+
+These same requirements exist when you use [`Function.toJS`] to make a Dart
+function callable in JS. The values that flow into and out of this callback must
+be a compatible interop type or a primitive.
+
+If you use a Dart primitive e.g. `String`, an implicit conversion happens in the
+compiler to convert that value from a JS value to a Dart value. If performance
+is critical and you don’t need to examine the contents of the string, then using
+`JSString` instead to avoid the conversion cost may make sense.
```

Comment:
````suggestion
`JSString` instead to avoid the conversion cost might make sense:

```dart
```
````

Would it make sense to show a small example of what you mean by using `JSString` instead? Or is that basically what you've already shown in the second `good` example?

Reply:
Yup! I mentioned the second example to be more explicit.